ProcessLib/SmallDeformation/SmallDeformationFEM.h

```diff
@@ -0,0 +1,356 @@
+            ip_data._b_matrices.resize(
+                KelvinVectorDimensions<DisplacementDim>::value,
+                ShapeFunction::NPOINTS * DisplacementDim);
+            LinearBMatrix::computeBMatrix<
+                DisplacementDim, ShapeFunction::NPOINTS,
+                typename ShapeMatricesType::GlobalDimNodalMatrixType,
+                BMatrixType>(shape_matrices[ip].dNdx, ip_data._b_matrices);
+
+            ip_data._sigma.resize(
+                KelvinVectorDimensions<DisplacementDim>::value);
+            ip_data._sigma_prev.resize(
+                KelvinVectorDimensions<DisplacementDim>::value);
+            ip_data._eps.resize(KelvinVectorDimensions<DisplacementDim>::value);
+            ip_data._eps_prev.resize(
+                KelvinVectorDimensions<DisplacementDim>::value);
+            ip_data._C.resize(KelvinVectorDimensions<DisplacementDim>::value,
+                              KelvinVectorDimensions<DisplacementDim>::value);
+
+            _secondary_data.N[ip] = shape_matrices[ip].N;
+        }
+    }
+
+    void assembleConcrete(
+        double const t, std::vector<double> const& local_x,
+        NumLib::LocalToGlobalIndexMap::RowColumnIndices const& indices,
+        GlobalMatrix& /*M*/, GlobalMatrix& /*K*/, GlobalVector& b) override
+    {
+        _localRhs.setZero();
+
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        NodalDisplacementVectorType local_displacement(ShapeFunction::NPOINTS *
+                                                       DisplacementDim);
+
+        for (int i = 0; i < ShapeFunction::NPOINTS * DisplacementDim; ++i)
+        {
+            local_displacement[i] = local_x[i];
+        }
+
+        SpatialPosition x_position;
+        x_position.setElementID(_element.getID());
+
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            x_position.setIntegrationPoint(ip);
+            auto const& wp = _integration_method.getWeightedPoint(ip);
+            auto const& detJ = _ip_data[ip]._detJ;
+
+            auto const& B = _ip_data[ip]._b_matrices;
+            auto const& eps_prev = _ip_data[ip]._eps_prev;
+            auto const& sigma_prev = _ip_data[ip]._sigma_prev;
+
+            auto& eps = _ip_data[ip]._eps;
+            auto& sigma = _ip_data[ip]._sigma;
+            auto& C = _ip_data[ip]._C;
+            auto& internal_variables = *_ip_data[ip]._internal_variables;
+
+            eps.noalias() = B * local_displacement;
+
+            _ip_data[ip]._solid_material.computeConstitutiveRelation(
+                t, x_position, _process_data.dt, eps_prev, eps, sigma_prev,
+                sigma, C, internal_variables);
+
+            _localRhs.noalias() -=
+                B.transpose() * sigma * detJ * wp.getWeight();
+
+            _secondary_data._sigmaXX[ip] = sigma[0];
+            _secondary_data._sigmaYY[ip] = sigma[1];
+            _secondary_data._sigmaXY[ip] = sigma[3];
```

Comment:
Note: Since you store sigma in the `_ip_data`, it is not necessary to save them separately only for extrapolation/output.